src/main/java/dev/jbang/JdkManager.java

```diff
@@ -0,0 +1,91 @@
+package dev.jbang;
+
+import static java.lang.System.getenv;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.Collectors;
+
+public class JdkManager {
+	private static final String JDK_DOWNLOAD_URL = "https://api.adoptopenjdk.net/v3/binary/latest/%d/ga/%s/%s/jdk/hotspot/normal/adoptopenjdk";
+
+	private static final int DEFAULT_JAVA_VERSION = 11;
+
+	public static Path getCurrentJdk(Integer expectedVersion) throws IOException {
+		Integer currentVersion = determineJavaVersion();
+		if (expectedVersion != null) {
+			if (currentVersion == null || currentVersion < expectedVersion) {
+				return getJdk(expectedVersion);
+			} else {
+				return getJdkHome();
+			}
+		} else {
+			if (currentVersion == null || currentVersion < 8) {
+				return getJdk(DEFAULT_JAVA_VERSION);
+			} else {
+				return getJdkHome();
+			}
+		}
+	}
+
+	public static Path getJdk(int version) throws IOException {
+		Path jdkDir = getJdkPath(version);
+		if (!Files.isDirectory(jdkDir)) {
+			jdkDir = downloadAndInstallJdk(version);
+		}
+		return jdkDir;
+	}
+
+	public static Path downloadAndInstallJdk(int version) throws IOException {
+		String url = String.format(JDK_DOWNLOAD_URL, version, Util.getOS().name(), Util.getArch().name());
+		Path jdkPkg = Util.downloadAndCacheFile(url, false);
+		Path jdkDir = getJdkPath(version);
+		UnpackUtil.unpack(jdkPkg, jdkDir);
+		return jdkDir;
+	}
+
+	private static Path getJdkPath(int version) {
+		return Settings.getCacheDir().resolve("jdks").resolve(Integer.toString(version));
+	}
+
+	private static Integer determineJavaVersion() {
```

Comment:
we got ModularClassPath.supportsModules() that does something similar - wanna check if we can unify/reuse this over there to not have conflicting/multiple calls to "java -version"

Reply:
Oh really? Didn't know that, never looked at that part of the code. Will do.